src/main/scala/chisel3/util/BitPat.scala

```diff
@@ -165,6 +166,8 @@ package experimental {
     /** whether this [[BitSet]] is empty (i.e. no value matches) */
     def isEmpty: Boolean = terms.forall(_.isEmpty)
 
+    def matches(input: UInt) = VecInit(terms.map(_ === input).toSeq).asUInt.orR
```

Comment:
Maybe override `BitSetRange.matches` to use `input >= start.U && input < (start + length).U`? It should leverage the power of synthesis to generate two adder, while `BitSet.matches` will generate a high fanout `OR` and a lot of `XOR` for misaligned `BitSetRange`.

Reply:
I believe this difference in behavior may surprise users? For example, `rangeA.matches(input)` generates two comparisons, while
```
rangeA.intersects(rangeB).matches(input)
```
generates a decoder, although the intersection is always a range itself.

I suggest that we leave `matches` as-is in BitSetRange, while adding a new function (e.g. `matchAsCmp`) to BitSetRange, so users will be alerted when they find out `rangeA.intersects(rangeB)` has no such member function, and no unexpected behavior will emerge.

Later, we may add APIs such as `BitSetRange.intersectRange(BitSetRange)` or `BitSet.tryAsRange`.

Reply:
Also, we may change matches into using something similar to:
```scala
object MatchAlgo extends Enumeration {
  val Auto, Cmp, Decoder = Value()
}
def matches(input: UInt, algo: MatchAlgo = MatchAlgo.Auto)
```

And allow using comparisons on all BitSets (See https://github.com/chipsalliance/chisel3/pull/2449#issuecomment-1075889265)

Reply:
Yes, this sounds to be a better idea.
IMO: 
1. `matches` for`BitPat` is only the `===` alias.
2. `BitSet` should use `Decoder`.
3. `BitSetRange` should use `Compare`(I prefer `Compare` than `Cmp` ;p)
If user explicitly uses the `Compare` option for `BitSet`, `BitSet` should call `bitSet.toBitSetRanges.map(_.matches(input)).reduce(_ || _)`.